### DISCUSSION TITLE
Pytorch Hub Example Does Not Detect Any Objects

### DISCUSSION THREAD

**Discussion by jritch** [ultralytics/yolov5#2839] [Q&A]
Hi, when I run the detailed example from https://github.com/ultralytics/yolov5/issues/36, I'm able to download and execute the model but it doesn't detect anything in the zidane or bus images (output tensors are empty). Does anyone have any idea as to why this might be or what I could do to try and fix it?

 ```
C:\Users\jacob\stanford> & C:/Python38/python.exe c:/Users/jacob/stanford/research/audio_ar/ml-model-training/pytorch-yolo.py
Using cache found in C:\Users\jacob/.cache\torch\hub\ultralytics_yolov5_master

                 from  n    params  module                                  arguments
  0                -1  1      3520  models.common.Focus                     [3, 32, 3]
  1                -1  1     18560  models.common.Conv                      [32, 64, 3, 2]
  2                -1  1     18816  models.common.C3                        [64, 64, 1]
  3                -1  1     73984  models.common.Conv                      [64, 128, 3, 2]
  4                -1  1    156928  models.common.C3                        [128, 128, 3]
  5                -1  1    295424  models.common.Conv                      [128, 256, 3, 2]
  6                -1  1    625152  models.common.C3                        [256, 256, 3]
  7                -1  1   1180672  models.common.Conv                      [256, 512, 3, 2]
  8                -1  1    656896  models.common.SPP                       [512, 512, [5, 9, 13]]
  9                -1  1   1182720  models.common.C3                        [512, 512, 1, False]
 10                -1  1    131584  models.common.Conv                      [512, 256, 1, 1]
 11                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']
 12           [-1, 6]  1         0  models.common.Concat                    [1]
 13                -1  1    361984  models.common.C3                        [512, 256, 1, False]
 14                -1  1     33024  models.common.Conv                      [256, 128, 1, 1]
 15                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']
 16           [-1, 4]  1         0  models.common.Concat                    [1]
 17                -1  1     90880  models.common.C3                        [256, 128, 1, False]
 18                -1  1    147712  models.common.Conv                      [128, 128, 3, 2]
 19          [-1, 14]  1         0  models.common.Concat                    [1]
 20                -1  1    296448  models.common.C3                        [256, 256, 1, False]
 21                -1  1    590336  models.common.Conv                      [256, 256, 3, 2]
 22          [-1, 10]  1         0  models.common.Concat                    [1]
 23                -1  1   1182720  models.common.C3                        [512, 512, 1, False]
 24      [17, 20, 23]  1    229245  models.yolo.Detect                      [80, [[10, 13, 16, 30, 33, 23], [30, 61, 62, 45, 59, 119], [116, 90, 156, 198, 373, 326]], [128, 256, 512]]
Model Summary: 283 layers, 7276605 parameters, 7276605 gradients, 17.1 GFLOPS

Adding autoShape... 
YOLOv5  2021-4-18 torch 1.8.1+cu111 CUDA:0 (GeForce GTX 1650, 4096.0MB)

Downloading zidane.jpg...
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 165k/165k [00:00<00:00, 6.22MB/s]
Downloading bus.jpg...
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476k/476k [00:00<00:00, 8.13MB/s]
image 1/2: 720x1280
image 2/2: 1080x810
Speed: 13.0ms pre-process, 952.1ms inference, 1.0ms NMS per image at shape (2, 3, 640, 640)
tensor([], device='cuda:0', size=(0, 6))
Empty DataFrame
Columns: [xmin, ymin, xmax, ymax, confidence, class, name]
Index: []
```

**Comment by jritch** [comment]
Ah, I see, it doesn't work on Windows when you use the GPU? https://github.com/ultralytics/yolov5/discussions/2713

I'd be interested in why this happens - can I still train using the GPU?

Thanks for this project, it's super helpful.

**Reply by glenn-jocher** [reply]
@jritch hi thanks for the bug report! I can't say for sure what's wrong as we don't test on Windows GPU instances, but Windows has a history of being problematic with PyTorch in general. I would recommend a Docker environment if you want to run YOLOv5 locally, or alternatively one of the cloud or notebook environments below, which are fully verified.

@SkalskiP @KalenMike this is an example of the CI challenges we discussed earlier.


## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab and Kaggle** notebooks with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/GCP-Quickstart)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/AWS-Quickstart)
- **Docker Image**. See [Docker Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/Docker-Quickstart) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>


## Status

![CI CPU testing](https://github.com/ultralytics/yolov5/workflows/CI%20CPU%20testing/badge.svg)

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are passing. These tests evaluate proper operation of basic YOLOv5 functionality, including training ([train.py](https://github.com/ultralytics/yolov5/blob/master/train.py)), testing ([test.py](https://github.com/ultralytics/yolov5/blob/master/test.py)), inference ([detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py)) and export ([export.py](https://github.com/ultralytics/yolov5/blob/master/models/export.py)) on MacOS, Windows, and Ubuntu.

**Reply by SkalskiP** [reply]
@glenn-jocher I understand. It is worth mentioning that CI/CD automation on Windows + GPU, can be very problematic. I don't even know if you can spin up any preconfigured windows server with GPU and PyTorch support on AWS. We would need to research that. Are we planning to support that combo? 

@jritch If you want to operate on Windows I would much rather recommend using WSL, than a plain Windows environment. From what I heard Nvidia support is not perfect there but it exists - [link](https://developer.nvidia.com/cuda/wsl). Let us know if you managed to set up the environment and run yolov5 in it.

**Reply by glenn-jocher** [reply]
@SkalskiP good points. Like you said I'm not even sure if it's possible to run CI on Windows GPU, or even if it were possible I don't know if the addressable market is worth the expense. There is certainly lower hanging fruit to pick first.

**Reply by SkalskiP** [reply]
@glenn-jocher my point exactly. Even if that would be possible I don't think it is worth the affort. btw: GH actions don't offer GPU support directly, however from my (small) research it should be possible to trigger GH action to run tests on external GPU.
